presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java

```diff
@@ -78,17 +85,118 @@ public static Slice regexpReplace(@SqlType(StandardTypes.VARCHAR) Slice source,
     @Description("replaces substrings matching a regular expression by given string")
     @ScalarFunction
     @SqlType(StandardTypes.VARCHAR)
-    public static Slice regexpReplace(@SqlType(StandardTypes.VARCHAR) Slice source, @SqlType(RegexpType.NAME) Pattern pattern, @SqlType(StandardTypes.VARCHAR) Slice replacement)
+    public static Slice regexpReplace(@SqlType(StandardTypes.VARCHAR) Slice source, @SqlType(RegexpType.NAME) Regex pattern, @SqlType(StandardTypes.VARCHAR) Slice replacement)
     {
-        Matcher matcher = pattern.matcher(source.toString(UTF_8));
-        String replaced = matcher.replaceAll(replacement.toString(UTF_8));
-        return Slices.copiedBuffer(replaced, UTF_8);
+        Matcher matcher = pattern.matcher(source.getBytes());
+        SliceOutput sliceOutput = new DynamicSliceOutput(source.length() + replacement.length() * 5);
+
+        int lastEnd = 0;
+        int nextStart = 0;
+        while (true) {
+            int offset = matcher.search(nextStart, source.length(), Option.DEFAULT);
+            if (offset == -1) {
+                break;
+            }
+            if (matcher.getEnd() == matcher.getBegin()) {
+                nextStart = matcher.getEnd() + 1;
+            }
+            else {
+                nextStart = matcher.getEnd();
+            }
+            Slice sliceBetweenReplacements = source.slice(lastEnd, matcher.getBegin() - lastEnd);
+            lastEnd = matcher.getEnd();
```

Comment:
I found it confusing that we don't set `lastEnd = nextStart`, add some comments to this loop